pcap/run.go

```diff
@@ -35,7 +36,9 @@ func Collect(stop <-chan struct{}, intf, bpfFilter string, bufferShare float32,
 
 	for t := range parsedChan {
 		t.Interface = intf
-		if err := proc.Process(t); err != nil {
+		err := proc.Process(t)
+		t.Content.ReleaseBuffers()
```

Comment:
Hm, I was misunderstanding this in my akita-libs review, there's no need to worry about different readers of the buffer having different lifetimes.

But, that does suggest that we maybe need not actually allocate the buffers from a fixed pool; we only need synchronize on how much memory is used.  I don't think this is a blocker; I don't want you to rewrite it.  But I do want to put these on the record in case we need to pursue further performance tuning.

In a multi-CPU system the allocator can be thread- or CPU-local which avoids thrashing cache lines.  With the shared pool we're giving up any potential optimization here.  (I know tcmalloc does exactly this sort of thing.  What the Go runtime does, I have no idea.)

Keeping the buffer live has a cost in footprint and in future profiling efforts.  It also turns an accounting issue (early release of the buffer before reading is done) into a correctness issue (re-use of old data.)

That said, if this is causing spikes then maybe re-use is a win because the GC isn't having to wake up.  So let's definitely keep it as-is, but maybe we want to consider making our buffers "virtual" rather than "real".

Reply:
Yeah, my primary goal was to have a memory pool to avoid GC altogether. Caching issues did occur to me, but I figured we could start with something simple first and dig deeper in the future, if desired.